src/main/java/com/epam/reportportal/auth/integration/saml/SamlUserReplicator.java

```diff
@@ -0,0 +1,146 @@
+        this.serverSettingsRepository = serverSettingsRepository;
+    }
+
+    public User replicateUser(ReportPortalSamlAuthentication samlAuthentication) {
+        String userName = normalizeId(samlAuthentication.getPrincipal());
+        User user = userRepository.findOne(userName);
+
+        if (user == null) {
+
+            ServerSettings serverSettings = serverSettingsRepository.findOne("default");
+            Map<String, SamlProviderDetails> samlProviders = Optional.ofNullable(serverSettings.getSamlProviderDetails()).orElseGet(HashMap::new);
+            Optional<SamlProviderDetails> first = samlProviders.values().stream()
+                    .filter(provider -> provider.getIdpUrl().equalsIgnoreCase(samlAuthentication.getIssuer()))
+                    .findFirst();
+
+            user = new User();
+            user.setLogin(userName);
+
+            List<Attribute> details = samlAuthentication.getDetails();
+
+            if (first.isPresent()) {
+                populateUserDetailsIfSettingsArePresent(user, first.get(), details);
+            } else {
+                populateUserDetails(user, details);
+            }
+
+            user.setType(UserType.SAML);
+            user.setRole(UserRole.USER);
+            user.setIsExpired(false);
+
+            user.setDefaultProject(generatePersonalProject(user));
+            user.setMetaInfo(defaultMetaInfo());
+
+            userRepository.save(user);
+        }
+
+        return user;
+    }
+
+    private void populateUserDetails(User user, List<Attribute> details) {
+        String email = findAttributeValue(details, UserAttribute.EMAIL.toString(), String.class);
+        checkEmail(email);
+        user.setEmail(email);
+
+        String firstName = findAttributeValue(details, UserAttribute.FIRST_NAME.toString(), String.class);
+        String lastName = findAttributeValue(details, UserAttribute.LAST_NAME.toString(), String.class);
+        user.setFullName(String.join(" ", firstName, lastName));
+    }
+
+    private void populateUserDetailsIfSettingsArePresent(User user, SamlProviderDetails providerDetails, List<Attribute> details) {
+        String email = findAttributeValue(details, providerDetails.getEmailAttributeId(), String.class);
+        checkEmail(email);
+        user.setEmail(email);
+
+        if (StringUtils.isEmpty(providerDetails.getFullNameAttributeId())) {
+            String firstName = findAttributeValue(details, providerDetails.getFirstNameAttributeId(), String.class);
+            String lastName = findAttributeValue(details, providerDetails.getLastNameAttributeId(), String.class);
+            user.setFullName(String.join(" ", firstName, lastName));
+        } else {
+            String fullName = findAttributeValue(details, providerDetails.getFullNameAttributeId(), String.class);
+            user.setFullName(fullName);
+        }
+    }
+
+    private <T> T findAttributeValue(List<Attribute> attributes, String lookingFor, Class<T> castTo) {
```

Comment:
some unit tests for this magic, please :)

Reply:
Added tests fir SamlUserReplicator